OmsAgent/Fairfax/omsagent_ff.py

```diff
@@ -430,6 +428,196 @@ def check_workspace_id_and_key(workspace_id, workspace_key):
+    """
+    If the VM already has a workspace/SCOM configured, then we should
+    disallow a new connection when stopOnMultipleConnections is used
+
+    Throw an exception in these cases:
+    - The workspace with the given workspace_id has not been onboarded
+      to the VM, but at least one other workspace has been
+    - The workspace with the given workspace_id has not been onboarded
+      to the VM, and the VM is connected to SCOM
+
+    If the extension operation is connecting to an already-configured
+    workspace, it is not a stopping case
+    """
+    other_connection_exists = False
+    if os.path.exists(OmsAdminPath):
+        exit_code, output = run_get_output(WorkspaceCheckCommand,
+                                           chk_err = False)
+
+        for line in output.split('\n'):
+            if workspace_id in line:
+                hutil_log_info('The workspace to be enabled has already ' \
+                               'been configured on the VM before; ' \
+                               'continuing despite stopOnMultipleConnections ' \
+                               'flag')
+                return
+            else:
+                # Note: if scom workspace dir is created, a line containing
+                # "Workspace(SCOM Workspace): scom" will be here
+                # If any other line is here, it may start sending data later
+                other_connection_exists = True
+    else:
+        for dir_name, sub_dirs, files in os.walk(EtcOmsAgentPath):
+            for sub_dir in sub_dirs:
+                sub_dir_name = os.path.basename(sub_dir)
+                workspace_search = re.compile(GUIDOnlyRegex, re.M)
+                if sub_dir_name == workspace_id:
+                    hutil_log_info('The workspace to be enabled has already ' \
+                                   'been configured on the VM before; ' \
+                                   'continuing despite ' \
+                                   'stopOnMultipleConnections flag')
+                    return
+                elif search.match(sub_dir_name) or sub_dir_name == 'scom':
+                    other_connection_exists = True
+
+    if other_connection_exists:
+        err_msg = ('This machine is already connected to some other Log ' \
+                   'Analytics workspace, please set ' \
+                   'stopOnMultipleConnections to false in public ' \
+                   'settings or remove this property, so this machine ' \
+                   'can connect to new workspaces, also it means this ' \
+                   'machine will get billed multiple times for each ' \
+                   'workspace it report to. ' \
+                   '(LINUXOMSAGENTEXTENSION_ERROR_MULTIPLECONNECTIONS)')
+        # This exception will get caught by the main method
+        raise OmsAgentUnwantedMultipleConnectionsException(err_msg)
+    else:
+        detect_scom_connection
+
+
+def detect_scom_connection():
```

Comment:
Seems very long to the function. Can you break it into multiple helper functions instead? Like "detect_omsadminpath", "detect_omiconfigeditor_path",etc.. and call them from detect_scom_connection()?

Reply:
Done